crates/client/sync/src/l2.rs

```diff
@@ -477,16 +486,40 @@ pub fn verify_l2<B: BlockT>(
     Ok(())
 }
 
-async fn update_highest_block_hash_and_number(client: &SequencerGatewayProvider) -> Result<(), String> {
-    let block = client.get_block(BlockId::Latest).await.map_err(|e| format!("failed to get block: {e}"))?;
-
-    let hash = block.block_hash.ok_or("block hash not found")?;
-    let number = block.block_number.ok_or("block number not found")?;
+async fn update_starknet_data<B, C>(provider: &SequencerGatewayProvider, client: &C) -> Result<(), String>
+where
+    B: BlockT,
+    C: HeaderBackend<B>,
+{
+    let block = provider.get_block(BlockId::Pending).await.map_err(|e| format!("Failed to get pending block: {e}"))?;
+
+    let hash_best = client.info().best_hash;
+    let hash_current = block.parent_block_hash;
+    // Well howdy, seems like we can't convert a B::Hash to a FieldElement pa'tner,
+    // fancy this instead? 🤠🔫
+    let tmp = <B as BlockT>::Hash::from_str(&hash_current.to_string()).unwrap_or(Default::default());
+    let number = block.block_number.ok_or("block number not found")? - 1;
+
+    // all blocks have been synchronized, can store pending data
+    if hash_best == tmp {
+        let state_update = provider
+            .get_state_update(BlockId::Pending)
+            .await
+            .map_err(|e| format!("Failed to get pending state update: {e}"))?;
+
+        // Speaking about type conversion hell: 🔥
+        *STARKNET_PENDING_BLOCK.write().expect("Failed to acquire write lock on STARKNET_PENDING_BLOCK") =
+            Some(crate::convert::block(block).await);
+
+        // This type conversion is evil and should not be necessary
+        *STARKNET_PENDING_STATE_UPDATE.write().expect("Failed to aquire write lock on STARKNET_PENDING_STATE_UPDATE") =
+            Some(crate::convert::state_update(state_update));
+    }
 
     let mut highest_block_hash_and_number = STARKNET_HIGHEST_BLOCK_HASH_AND_NUMBER
         .write()
         .expect("Failed to acquire write lock on STARKNET_HIGHEST_BLOCK_HASH_AND_NUMBER");
-    *highest_block_hash_and_number = (hash, number);
+    *highest_block_hash_and_number = (hash_current, number);
```

Comment:
refactor it like the other 2 RwLock, without using a mutable variable

Reply:
Will update for the final pr :+1:

Reply:
done!